### PR TITLE
Caller argument expression

### LINKED ISSUE #25680: Caller argument expression
[Feature proposal](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-10.0/caller-argument-expression.md)

This would aid in debugging and diagnostic scenarios. It would allow methods to capture the expression passed to a method for diagnostic purposes.

This should be updated in a few places:

- Where diagnostics are covered
- Attributes understood by the compiler.
- nameof expressions should point to this new feature.

### PR DESCRIPTION
**Pull request by BillWagner**
Fixes #25680 

This adds a description of the `CallerArgumentExpression` in the language reference.

### CODE REVIEW

**Review by 333fred — APPROVED** [review]

File: docs/csharp/language-reference/attributes/caller-information.md [hunk 1]
```diff
@@ -47,19 +26,51 @@ You specify an explicit default value for each optional parameter. You can't app
...
+| Constructor | The string ".ctor" |
+| Static constructor | The string ".cctor" |
+| Finalizer | The string "Finalize" |
+| User-defined operators or conversions | The generated name for the member, for example, "op_Addition". |
+| Attribute constructor | The name of the method or property to which the attribute is applied. If the attribute is any element within a member (such as a parameter, a return value, or a generic type parameter), this result is the name of the member that's associated with that element. |
+| No containing member (for example, assembly-level or attributes that are applied to types) | The default value of the optional parameter. |
+
+## Argument expressions
+
+You use the <xref:System.Runtime.CompilerServices.CallerArgumentExpressionAttribute?displayProperty=nameWithType> when you want the expression used for an argument. Diagnostic libraries may want to provide more details about the *expressions* used for arguments. By providing the expression that triggered the diagnostic, in addition to the parameter name, developers see more details about the code that triggered the diagnostic. That extra information makes it easier to fix.
```
Comment (KathleenDollard) [inline docs/csharp/language-reference/attributes/caller-information.md:45]: Consider:

```suggestion
You use the <xref:System.Runtime.CompilerServices.CallerArgumentExpressionAttribute?displayProperty=nameWithType> when you want the expression used for an argument. Diagnostic libraries may want to provide more details about the *expressions* passed to arguments. By providing the expression that triggered the diagnostic, in addition to the parameter name, developers see more details about the code that triggered the diagnostic. That extra information makes it easier to fix.
```

File: docs/csharp/language-reference/attributes/caller-information.md [hunk 2]
```diff
@@ -47,19 +26,51 @@ You specify an explicit default value for each optional parameter. You can't app
...
+
+The following example shows how you can provide detailed information about the argument when it's invalid:
+
+:::code language="csharp" source="./snippets/CallerInformation.cs" id="TestCondition":::
+
+You would invoke it as shown in the following example:
+
+:::code language="csharp" source="./snippets/CallerInformation.cs" id="InvokeTestCondition":::
+
+The expression used for `condition` is injected by the compiler into the `message` argument. When a developer calls `Operation` with a `null` argument, the following message is stored in the `ArgumentException`:
```
Comment (KathleenDollard) [inline docs/csharp/language-reference/attributes/caller-information.md:55]: Consider rewording, and in your code InvalidOp not Argument looks like its thrown. 

```suggestion
The expression used for `condition` is injected by the compiler into the `message` argument. When a developer calls `Operation` with a `null` argument, an `InvalidOperationException` is thrown with following message: 
```

File: docs/csharp/whats-new/csharp-10.md [hunk 3]
```diff
@@ -117,3 +117,19 @@ int x = 0;
 In C# 10.0 and later, you can specify a different async method builder for a single method, in addition to specifying the method builder type for all methods that return a given task-like type. A custom async method builder enables advanced performance tuning scenarios where a given method may benefit from a custom builder.
 
 To learn more, see the section on [`AsyncMethodBuilder`](../language-reference/attributes/general.md#asyncmethodbuilder-attribute) in the article on attributes read by the compiler.
+
+## CallerArgumentExpression attribute diagnostics
+
+You can use the <xref:System.Runtime.CompilerServices.CallerArgumentExpressionAttribute?displayProperty=fullName> to specify a parameter that the compiler replaces with the text representation of another argument. This feature enables libraries to create more specific diagnostics. The following code tests a condition. If the condition is false, the exception message contains the text representation of the argument for `condition`:
```
Comment (KathleenDollard) [inline docs/csharp/whats-new/csharp-10.md:123]: Consider:
```suggestion
You can use the <xref:System.Runtime.CompilerServices.CallerArgumentExpressionAttribute?displayProperty=fullName> to specify a parameter that the compiler replaces with the text representation of another argument. This feature enables libraries to create more specific diagnostics. The following code tests a condition. If the condition is false, the exception message contains the text representation of the argument passed to `condition`:
```

File: docs/csharp/whats-new/csharp-10.md [hunk 4]
```diff
@@ -117,3 +117,19 @@ int x = 0;
...
+public static void Validate(bool condition, [CallerArgumentExpression("condition")] string? message=null)
+{
+    if (!condition)
+    {
+        throw new InvalidOperationException($"Argument failed validation: <{message}>");
+    }
+}
+```
+
+You can learn more about this feature in the article on [Caller information attributes](../language-reference/attributes/caller-information.md#argument-expressions) in the language reference section.
```
Comment (KathleenDollard) [inline docs/csharp/whats-new/csharp-10.md:135]: I think this is hte article I flagged in an issue as needing review. Or maybe I didn't save that as you said you were working on this.

File: docs/visual-basic/programming-guide/concepts/caller-information.md [hunk 5]
```diff
@@ -73,6 +74,19 @@ End Sub
 |Attribute constructor|The name of the member to which the attribute is applied. If the attribute is any element within a member (such as a parameter, a return value, or a generic type parameter), this result is the name of the member that's associated with that element.|  
 |No containing member (for example, assembly-level or attributes that are applied to types)|The default value of the optional parameter.|  
   
+### <a name="CALLEREXPRESSION"></a> Caller expressions  
+
+You use the <xref:System.Runtime.CompilerServices.CallerArgumentExpressionAttribute?displayProperty=nameWithType> when you want the expression used for an argument. Diagnostic libraries may want to provide more details about the *expressions* used for arguments. By providing the expression that triggered the diagnostic, in addition to the parameter name, developers see more details about the code that triggered the diagnostic. That extra information makes it easier to fix. The following method uses the <xref:System.Runtime.CompilerServices.CallerArgumentExpressionAttribute> to display the condition that must be `True`:
```
Comment (KathleenDollard) [inline docs/visual-basic/programming-guide/concepts/caller-information.md:79]: Again, I'm not sure users will relate to "used for" as much as "passed..." I think folks think of passing arguments around before arriving at a method, in which they are used.

**Review by KathleenDollard — COMMENTED** [review]
Looks good other than one mistake in an exception name. I also had opinions.

File: docs/csharp/language-reference/attributes/caller-information.md [hunk 6]
```diff
@@ -13,53 +13,64 @@ Using info attributes, you obtain information about the caller to a method. You
 |<xref:System.Runtime.CompilerServices.CallerFilePathAttribute>|Full path of the source file that contains the caller. The full path is the path at compile time.|`String`|
 |<xref:System.Runtime.CompilerServices.CallerLineNumberAttribute>|Line number in the source file from which the method is called.|`Integer`|
 |<xref:System.Runtime.CompilerServices.CallerMemberNameAttribute>|Method name or property name of the caller.|`String`|
+| <xref:System.Runtime.CompilerServices.CallerArgumentExpressionAttribute>|String representation of the argument expression.|
```
Comment (gewarren) [inline docs/csharp/language-reference/attributes/caller-information.md:16]:
```suggestion
| <xref:System.Runtime.CompilerServices.CallerArgumentExpressionAttribute>|String representation of the argument expression.|`String`|
```

File: docs/csharp/language-reference/attributes/caller-information.md [hunk 7]
```diff
@@ -13,53 +13,64 @@ Using info attributes, you obtain information about the caller to a method. You
...
-    Trace.WriteLine("source line number: " + sourceLineNumber);
-}
-
-// Sample Output:
-//  message: Something happened.
-//  member name: DoProcessing
-//  source file path: c:\Visual Studio Projects\CallerInfoCS\CallerInfoCS\Form1.cs
-//  source line number: 31
-```
+This information helps you write tracing, debugging, and create diagnostic tools. The following example shows how to use caller info attributes. On each call to the `TraceMessage` method, the caller information is inserted for the arguments to the optional parameters.
```
Comment (gewarren) [inline docs/csharp/language-reference/attributes/caller-information.md:18]:
```suggestion
This information helps you with tracing and debugging, and helps you to create diagnostic tools. The following example shows how to use caller info attributes. On each call to the `TraceMessage` method, the caller information is inserted for the arguments to the optional parameters.
```

File: docs/csharp/language-reference/attributes/caller-information.md (hunk 1 above)
Comment (gewarren) [reply docs/csharp/language-reference/attributes/caller-information.md:45]:
```suggestion
You use the <xref:System.Runtime.CompilerServices.CallerArgumentExpressionAttribute?displayProperty=nameWithType> when you want the expression used for an argument. Diagnostic libraries may want to provide more details about the *expressions* used for arguments. By providing the expression that triggered the diagnostic, in addition to the parameter name, developers have more details about the condition that triggered the diagnostic. That extra information makes it easier to fix.
```

File: docs/csharp/whats-new/csharp-10.md [hunk 8]
```diff
@@ -117,3 +117,19 @@ int x = 0;
...
+## CallerArgumentExpression attribute diagnostics
+
+You can use the <xref:System.Runtime.CompilerServices.CallerArgumentExpressionAttribute?displayProperty=fullName> to specify a parameter that the compiler replaces with the text representation of another argument. This feature enables libraries to create more specific diagnostics. The following code tests a condition. If the condition is false, the exception message contains the text representation of the argument for `condition`:
+
+```csharp
+public static void Validate(bool condition, [CallerArgumentExpression("condition")] string? message=null)
+{
+    if (!condition)
+    {
+        throw new InvalidOperationException($"Argument failed validation: <{message}>");
```
Comment (gewarren) [inline docs/csharp/whats-new/csharp-10.md:130]:
```suggestion
        throw new InvalidOperationException($"Expression failed validation: <{message}>");
```

File: docs/visual-basic/programming-guide/concepts/caller-information.md (hunk 5 above)
Comment (gewarren) [inline docs/visual-basic/programming-guide/concepts/caller-information.md:79]:
```suggestion
You use the <xref:System.Runtime.CompilerServices.CallerArgumentExpressionAttribute?displayProperty=nameWithType> when you want to capture the expression used for an argument. Diagnostic libraries may want to provide more details about the *expressions* used for arguments. By providing the expression that triggered the diagnostic, in addition to the parameter name, developers see more details about the condition that triggered the diagnostic. That extra information makes it easier to fix. The following method uses the <xref:System.Runtime.CompilerServices.CallerArgumentExpressionAttribute> to display the condition that must be `True`:
```

**Review by gewarren — COMMENTED** [review]
I left some suggestions for consideration.

File: docs/csharp/whats-new/csharp-10.md (hunk 8 above)
Comment (BillWagner) [reply docs/csharp/whats-new/csharp-10.md:130]: I changed to ArgumentException, so left the message the same.

File: docs/csharp/whats-new/csharp-10.md (hunk 4 above)
Comment (gewarren) [reply docs/csharp/whats-new/csharp-10.md:135]: @BillWagner I think you meant https://github.com/dotnet/docs/issues/26784. Can you confirm that the attribute is implemented in the compiler now? If so, we'll need to manually remove the note block.

Comment (BillWagner) [reply docs/csharp/whats-new/csharp-10.md:135]: Yes I did. Thanks @gewarren 

Yes, this is implemented in the compiler now, as of the current public RC for .NET 6.

Comment (gewarren) [reply docs/csharp/whats-new/csharp-10.md:135]: Updated here: https://github.com/dotnet/dotnet-api-docs/pull/7351/files

**Comment by cremor** [comment]
@BillWagner On the top of the "What's new in C# 10.0" page there is a list of items. It looks like a table of contents but isn't. You didn't add a link to your new chapter there, was this on purpose?

**Pull request merged**